registry/auth/token/accesscontroller_test.go

```diff
@@ -0,0 +1,36 @@
+package token
+
+import (
+	"crypto/tls"
+	"net/http/httptest"
+	"testing"
+)
+
+func TestBuildAutoRedirectURL(t *testing.T) {
+	t.Run("httpRequest", func(t *testing.T) {
+		req := httptest.NewRequest("GET", "http://example.com/", nil)
+		expectedURL := "http://example.com/auth/token"
+		actualURL := buildAutoRedirectURL(req)
+		assertURLEqual(t, expectedURL, actualURL)
+	})
+	t.Run("httpsRequest", func(t *testing.T) {
+		req := httptest.NewRequest("GET", "https://example.com/", nil)
+		req.TLS = new(tls.ConnectionState)
+		expectedURL := "https://example.com/auth/token"
+		actualURL := buildAutoRedirectURL(req)
+		assertURLEqual(t, expectedURL, actualURL)
+	})
+	t.Run("xforwarded", func(t *testing.T) {
+		req := httptest.NewRequest("GET", "http://example.com/", nil)
+		req.Header.Set("X-Forwarded-Proto", "https")
+		expectedURL := "https://example.com/auth/token"
+		actualURL := buildAutoRedirectURL(req)
+		assertURLEqual(t, expectedURL, actualURL)
+	})
+}
+
+func assertURLEqual(t *testing.T, expected, actual string) {
+	if expected != actual {
+		t.Errorf("URL mismatch: expected %q, got %q", expected, actual)
+	}
+}
```

Comment:
Test helper functions should call `t.Helper()` so the caller's file:line gets logged on error, not the helper's. [And the error message is backwards.](https://go.dev/wiki/TestComments#got-before-want)

Making this helper just a bit more specialized would make a big impact on DRYing out the test cases.
```suggestion
func assertAutoRedirectURL(t *testing.T, req *http.Request, expected string) {
	t.Helper()
	if actual := buildAutoRedirectURL(req); expected != actual {
		t.Errorf("buildAutoRedirectURL(_) = %q; want %q", actual, expected)
	}
}
```

```go
t.Run("xforwarded", func(t *testing.T) {
	req := httptest.NewRequest("GET", "http://example.com/", nil)
	req.Header.Set("X-Forwarded-Proto", "https")
	assertAutoRedirectURL(t, req, "https://example.com/auth/token")
})

Reply:
Fixed, thanks.